lightning/src/offers/invoice_request.rs

```diff
@@ -0,0 +1,1115 @@
+		self.invoice_request.amount_msats = Some(amount_msats);
+		self
+	}
+
+	/// Sets the [`InvoiceRequest::features`].
+	///
+	/// Successive calls to this method will override the previous setting.
+	#[cfg(test)]
+	pub fn features(mut self, features: InvoiceRequestFeatures) -> Self {
+		self.invoice_request.features = features;
+		self
+	}
+
+	/// Sets [`InvoiceRequest::quantity`] of items. If not set, `1` is assumed. Must conform to
+	/// [`Offer::is_valid_quantity`].
+	///
+	/// Successive calls to this method will override the previous setting.
+	pub fn quantity(mut self, quantity: u64) -> Result<Self, SemanticError> {
+		self.invoice_request.offer.check_quantity(Some(quantity))?;
+		self.invoice_request.quantity = Some(quantity);
+		Ok(self)
+	}
+
+	#[cfg(test)]
+	pub fn quantity_unchecked(mut self, quantity: u64) -> Self {
+		self.invoice_request.quantity = Some(quantity);
+		self
+	}
+
+	/// Sets the [`InvoiceRequest::payer_note`].
+	///
+	/// Successive calls to this method will override the previous setting.
+	pub fn payer_note(mut self, payer_note: String) -> Self {
+		self.invoice_request.payer_note = Some(payer_note);
+		self
+	}
+
+	/// Builds an [`InvoiceRequest`] after checking for valid semantics.
+	pub fn build(mut self) -> Result<UnsignedInvoiceRequest<'a>, SemanticError> {
+		let chain = self.invoice_request.chain();
+		if !self.offer.supports_chain(chain) {
+			return Err(SemanticError::UnsupportedChain);
+		}
+
+		if chain == self.offer.implied_chain() {
+			self.invoice_request.chain = None;
+		}
+
+		if self.offer.amount().is_none() && self.invoice_request.amount_msats.is_none() {
+			return Err(SemanticError::MissingAmount);
+		}
+
+		if let Some(Amount::Currency { .. }) = self.offer.amount() {
+			return Err(SemanticError::UnsupportedCurrency);
+		}
+
+		self.invoice_request.offer.check_quantity(self.invoice_request.quantity)?;
+
+		if self.offer.features().requires_unknown_bits() {
+			return Err(SemanticError::UnknownRequiredFeatures);
+		}
+
+		let amount_msats = self.invoice_request.amount_msats.unwrap_or(self.offer.amount_msats());
+		let quantity = self.invoice_request.quantity.unwrap_or(1);
+		if amount_msats < self.offer.expected_invoice_amount_msats(quantity) {
+			return Err(SemanticError::InsufficientAmount);
+		}
+
+		let InvoiceRequestBuilder { offer, invoice_request } = self;
+		Ok(UnsignedInvoiceRequest { offer, invoice_request })
+	}
+
+	#[cfg(test)]
+	pub(super) fn build_unchecked(self) -> UnsignedInvoiceRequest<'a> {
+		let InvoiceRequestBuilder { offer, invoice_request } = self;
+		UnsignedInvoiceRequest { offer, invoice_request }
+	}
+}
+
+/// A semantically valid [`InvoiceRequest`] that hasn't been signed.
+pub struct UnsignedInvoiceRequest<'a> {
+	offer: &'a Offer,
+	invoice_request: InvoiceRequestContents,
+}
+
+impl<'a> UnsignedInvoiceRequest<'a> {
+	/// Signs the invoice request using the given function.
+	pub fn sign<F>(self, sign: F) -> Result<InvoiceRequest, secp256k1::Error>
```

Comment:
Will we have a utility that takes a `KeysManager` in the future?

Reply:
Yeah, that's the plan.